File: codeforces/1901/D.cpp
/***************************************************/
/*            Author : Md. Arik Rayhan             */
/*        Github : github.com/mdarikrayhan         */
/***************************************************/
#include <bits/stdc++.h>
using namespace std;
const char nl = '\n';

#define ll unsigned long long int
#define vl vector<ll>
#define rep(i, a, b) for (ll i = a; i < b; i++)

const int N = 3e5 + 5;
int a[N], pre[N], suf[N];

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    int n;
    cin >> n;
    for (int i = 1; i <= n; ++i)
    {
        cin >> a[i];
        pre[i] = max(pre[i - 1], a[i] + n - i);
    }

    for (int i = n; i >= 1; --i)
    {
        suf[i] = max(suf[i + 1], a[i] + i - 1);
    }
    int ans = 2e9 + 100;
    for (int i = 1; i <= n; ++i)
    {
        ans = min(ans, max({a[i], pre[i - 1], suf[i + 1]}));
    }
    cout << ans << "\n";
    return 0;
}

File: codeforces/1901/D.py
n=int(input())
arr = list(map(int, input().split()))
a = [0]*n
b = [0]*n
for i in range(1,n):
    a[i] = max(a[i-1], arr[i-1] + n - i)
    b[n-i-1] = max(b[n-i], arr[n-i] + n - i)
print(min(max(arr[i],a[i],b[i]) for i in range(n)))